File: Chapter 07/Python/section_7_4_event_code/infrastructure/in_memory_broker.py
# infrastructure/in_memory_broker.py

import asyncio
from ..core.event_publisher import EventPublisher
from ..shared.event import Event
from ..shared.order_placed import OrderPlaced
from ..handlers.shipping_label_printer import ShippingLabelPrinter

class InMemoryBroker(EventPublisher):
    """
    THE SHOCK ABSORBER: Simulates a real message broker using an asyncio.Queue.
    
    The queue safely holds messages in memory until a background consumer is 
    ready to pull them, protecting the system from sudden spikes in traffic.
    """
    def __init__(self):
        # asyncio.Queue is a thread-safe, async-native way to pass data 
        # between different tasks in the Python event loop.
        self.queue = asyncio.Queue()
        self.shipping_service = ShippingLabelPrinter()

    async def publish_async(self, event: Event) -> None:
        """
        Accepts a message and immediately returns control to the caller.
        """
        await self.queue.put(event)

    async def start_listening_async(self) -> None:
        """
        Simulates independent background workers pulling messages off the queue.
        This runs continuously in an infinite loop, completely separated from 
        the user-facing web API.
        """
        while True:
            try:
                # This will pause the loop until a message is placed in the queue
                event = await self.queue.get()
                
                # Routing logic: Ensure the right worker gets the right message
                if isinstance(event, OrderPlaced):
                    await self.shipping_service.handle_async(event)
                
                # Signal to the queue that the item has been fully processed
                self.queue.task_done()
                
            except asyncio.CancelledError:
                # This catches the cleanup signal from our Demo to shut down safely
                print("[Broker] Listener shutdown signal received. Memory safely reclaimed.")
                break

File: Chapter 07/Java/src/main/java/com/grokkingsoftwarearchitecture/chapter07/section_7_4_event_code/infrastructure/InMemoryBroker.java
package com.grokkingsoftwarearchitecture.chapter07.section_7_4_event_code.infrastructure;

import com.grokkingsoftwarearchitecture.chapter07.section_7_4_event_code.core.EventPublisher;
import com.grokkingsoftwarearchitecture.chapter07.section_7_4_event_code.shared.Event;
import com.grokkingsoftwarearchitecture.chapter07.section_7_4_event_code.shared.OrderPlaced;
import com.grokkingsoftwarearchitecture.chapter07.section_7_4_event_code.handlers.ShippingLabelPrinter;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.CompletableFuture;

/**
 * THE SHOCK ABSORBER: Simulates a real message broker (like RabbitMQ or Azure Service Bus).
 * This infrastructure layer completely decouples the Producer from the Consumer.
 */
public class InMemoryBroker implements EventPublisher {
    
    // We use Java's BlockingQueue to act as an in-memory queue. 
    // It safely holds messages in memory until a background consumer asks for them.
    private final BlockingQueue<Event> _channel = new LinkedBlockingQueue<>();
    private final ShippingLabelPrinter _shippingService = new ShippingLabelPrinter();

    /**
     * Accepts a message and immediately returns control to the caller.
     */
    @Override
    public <T extends Event> void publishAsync(T event) {
        _channel.offer(event);
    }

    /**
     * Simulates the background worker continuously pulling new messages as they arrive,
     * completely independent of the web request lifecycle.
     */
    public void startListeningAsync() {
        CompletableFuture.runAsync(() -> {
            try {
                while (true) {
                    Event event = _channel.take();
                    if (event instanceof OrderPlaced orderPlaced) {
                        _shippingService.handleAsync(orderPlaced).join();
                    }
                }
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
            }
        });
    }
}